store/store_download.go

```diff
@@ -80,14 +80,63 @@ func init() {
 }
 
 // Deltas enabled by default on classic, but allow opting in or out on both classic and core.
-func useDeltas() bool {
-	// only xdelta3 is supported for now, so check the binary exists here
-	// TODO: have a per-format checker instead
-	if _, err := getXdelta3Cmd(); err != nil {
+func (s *Store) useDeltas() (use bool) {
+	s.xdeltaCheckLock.Lock()
+	defer s.xdeltaCheckLock.Unlock()
+
+	// check the cached value if available
+	if s.shouldUseDeltas != nil {
+		return *s.shouldUseDeltas
+	}
+
+	defer func() {
+		// cache whatever value we return for next time
+		s.shouldUseDeltas = &use
+	}()
+
+	// check if deltas were disabled by the environment
+	if !osutil.GetenvBool("SNAPD_USE_DELTAS_EXPERIMENTAL", true) {
+		// then the env var is explicitly false, we can't use deltas
+		logger.Debugf("delta usage disabled by environment variable")
+		return false
+	}
+
+	// TODO: have a per-format checker instead, we currently only support
+	// xdelta3 as a format for deltas
+
+	// check if the xdelta3 config command works from the system snap
+	cmd, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3", "config")
+	if err == nil {
+		// we have a xdelta3 from the system snap, make sure it works
+		if runErr := cmd.Run(); runErr == nil {
```

Comment:
That's fine but I wonder if we should be paranoid over tools we ship in system snap. They should just work, no (if not, we probably have a bigger problem?).

Reply:
well the fact that we cache the result of this means that there isn't a performance penalty to this paranoia so I think we should stick with more paranoia, especially given the fact that it sort of changed under our hood and that how we run things from the system snaps using the interpreter manually is a bit weird and could be fragile long term

Reply:
Yeah, you have a point there, fair enough.